File: avm_project/ios_app/Loan4U_iOS/Services/MLModelService.swift
import Foundation
import onnxruntime_objc

struct PredictionResult: Equatable {
    let predictedPrice: Int
    let confidenceScore: Double
    let region: String
    let timestamp: Date
}

/// Loads the shipped ONNX models and runs inference via ONNX Runtime.
/// Both platforms load the identical `.onnx` files — no CoreML/TFLite
/// conversion step, so what the Python contract test verifies is what runs.
@MainActor
final class MLModelService: ObservableObject {
    static let shared = MLModelService()

    @Published private(set) var isLoaded = false
    @Published private(set) var loadingError: String?

    private var env: ORTEnv?
    private var sessions: [String: ORTSession] = [:]
    private let modelKeys = ["nationwide", "seoul", "busan", "gyeonggi", "daegu", "incheon"]

    func load() async {
        do {
            let env = try ORTEnv(loggingLevel: .warning)
            self.env = env
            var loaded: [String: ORTSession] = [:]
            for key in modelKeys {
                guard let url = Bundle.main.url(forResource: "KR_\(key)_lite", withExtension: "onnx") else {
                    continue
                }
                loaded[key] = try ORTSession(env: env, modelPath: url.path, sessionOptions: nil)
            }
            guard loaded["nationwide"] != nil else {
                throw NSError(domain: "Loan4U", code: 1,
                              userInfo: [NSLocalizedDescriptionKey: "nationwide model missing"])
            }
            sessions = loaded
            isLoaded = true
        } catch {
            loadingError = "Model load failed: \(error.localizedDescription)"
        }
    }

    /// Region-specific model when available, else the nationwide fallback.
    func predict(_ input: PropertyInput, region: String?) -> PredictionResult? {
        let key = region?.lowercased() ?? "nationwide"
        guard let session = sessions[key] ?? sessions["nationwide"] else { return nil }

        let vector = FeatureEngineering.buildVector(input)
        do {
            let price = try runInference(session: session, vector: vector)
            guard price > 0 else { return nil }
            return PredictionResult(
                predictedPrice: Int(price.rounded()),
                confidenceScore: confidence(forRegion: key),
                region: region ?? "Nationwide",
                timestamp: Date()
            )
        } catch {
            loadingError = "Inference failed: \(error.localizedDescription)"
            return nil
        }
    }

    private func runInference(session: ORTSession, vector: [Float]) throws -> Float {
        let data = NSMutableData(bytes: vector, length: vector.count * MemoryLayout<Float>.stride)
        let input = try ORTValue(
            tensorData: data,
            elementType: .float,
            shape: [1, NSNumber(value: vector.count)]
        )
        let outputs = try session.run(
            withInputs: ["float_input": input],
            outputNames: ["variable"],
            runOptions: nil
        )
        guard let value = outputs["variable"],
              let tensorData = try value.tensorData() as Data? else {
            throw NSError(domain: "Loan4U", code: 2)
        }
        return tensorData.withUnsafeBytes { $0.load(as: Float.self) }
    }

    /// Confidence reflects each region model's validated MAPE, not a
    /// per-prediction uncertainty (the tree models don't expose one).
    private func confidence(forRegion key: String) -> Double {
        let mape: [String: Double] = [
            "seoul": 0.0947, "busan": 0.0948, "gyeonggi": 0.0994,
            "daegu": 0.1094, "incheon": 0.0950, "nationwide": 0.8148,
        ]
        return max(0.0, min(1.0, 1.0 - (mape[key] ?? 0.5)))
    }
}

File: avm_project/android_app/Loan4U_Android/app/src/main/kotlin/com/loan4u/services/MLModelService.kt
package com.loan4u.services

import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import android.content.Context
import com.loan4u.features.FeatureEngineering
import com.loan4u.features.PropertyInput
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.nio.FloatBuffer
import javax.inject.Inject
import javax.inject.Singleton

data class PredictionResult(
    val predictedPrice: Int,
    val confidenceScore: Double,
    val region: String,
    val timestamp: Long,
)

/**
 * Loads the shipped ONNX models with ONNX Runtime for Android. Both platforms
 * load the identical `.onnx` files verified by the Python contract test — no
 * TFLite/CoreML conversion step.
 */
@Singleton
class MLModelService @Inject constructor(
    @ApplicationContext private val context: Context,
) {
    @Volatile var isLoaded: Boolean = false
        private set
    @Volatile var loadingError: String? = null
        private set

    private var env: OrtEnvironment? = null
    private val sessions = mutableMapOf<String, OrtSession>()
    private val modelKeys = listOf("nationwide", "seoul", "busan", "gyeonggi", "daegu", "incheon")

    private val regionMape = mapOf(
        "seoul" to 0.0947, "busan" to 0.0948, "gyeonggi" to 0.0994,
        "daegu" to 0.1094, "incheon" to 0.0950, "nationwide" to 0.8148,
    )

    /** Loads all models off the main thread. Safe to call once at startup. */
    suspend fun load() = withContext(Dispatchers.IO) {
        try {
            val environment = OrtEnvironment.getEnvironment()
            env = environment
            modelKeys.forEach { key ->
                context.assets.open("models/KR_${key}_lite.onnx").use { stream ->
                    val bytes = stream.readBytes()
                    sessions[key] = environment.createSession(bytes, OrtSession.SessionOptions())
                }
            }
            check(sessions.containsKey("nationwide")) { "nationwide model missing" }
            isLoaded = true
        } catch (e: Exception) {
            loadingError = "Model load failed: ${e.message}"
        }
    }

    fun predict(input: PropertyInput, region: String?): PredictionResult? {
        val key = region?.lowercase() ?: "nationwide"
        val session = sessions[key] ?: sessions["nationwide"] ?: return null
        val environment = env ?: return null

        return try {
            val vector = FeatureEngineering.buildVector(input)
            val tensor = OnnxTensor.createTensor(
                environment,
                FloatBuffer.wrap(vector),
                longArrayOf(1, vector.size.toLong()),
            )
            tensor.use { t ->
                session.run(mapOf("float_input" to t)).use { output ->
                    @Suppress("UNCHECKED_CAST")
                    val raw = (output[0].value as Array<FloatArray>)[0][0]
                    if (raw <= 0f) return null
                    PredictionResult(
                        predictedPrice = raw.toInt(),
                        confidenceScore = (1.0 - (regionMape[key] ?: 0.5)).coerceIn(0.0, 1.0),
                        region = region ?: "Nationwide",
                        timestamp = System.currentTimeMillis(),
                    )
                }
            }
        } catch (e: Exception) {
            loadingError = "Inference failed: ${e.message}"
            null
        }
    }
}
